src/state/farms/elastic/hooks.ts

```diff
@@ -0,0 +1,769 @@
+
+      return tx.hash
+    },
+    [addTransactionWithType, contract],
+  )
+
+  const emergencyWithdraw = useCallback(
+    async (nftIds: BigNumber[]) => {
+      if (!contract) {
+        throw new Error(CONTRACT_NOT_FOUND_MSG)
+      }
+      const estimateGas = await contract.estimateGas.emergencyWithdraw(nftIds)
+      const tx = await contract.emergencyWithdraw(nftIds, {
+        gasLimit: calculateGasMargin(estimateGas),
+      })
+      addTransactionWithType(tx, { type: 'ForceWithdraw' })
+
+      return tx.hash
+    },
+    [addTransactionWithType, contract],
+  )
+
+  const stake = useCallback(
+    async (pid: BigNumber, nftIds: BigNumber[], liqs: BigNumber[]) => {
+      if (!contract) {
+        throw new Error(CONTRACT_NOT_FOUND_MSG)
+      }
+
+      const estimateGas = await contract.estimateGas.join(pid, nftIds, liqs)
+      const tx = await contract.join(pid, nftIds, liqs, {
+        gasLimit: calculateGasMargin(estimateGas),
+      })
+      addTransactionWithType(tx, { type: 'Stake', summary: `liquidity into farm` })
+
+      return tx.hash
+    },
+    [addTransactionWithType, contract],
+  )
+
+  const unstake = useCallback(
+    async (pid: BigNumber, nftIds: BigNumber[], liqs: BigNumber[]) => {
+      if (!contract) {
+        throw new Error(CONTRACT_NOT_FOUND_MSG)
+      }
+      try {
+        const estimateGas = await contract.estimateGas.exit(pid, nftIds, liqs)
+        const tx = await contract.exit(pid, nftIds, liqs, {
+          gasLimit: calculateGasMargin(estimateGas),
+        })
+        addTransactionWithType(tx, { type: 'Unstake', summary: `liquidity from farm` })
+
+        return tx.hash
+      } catch (e) {
+        console.log(e)
+      }
+    },
+    [addTransactionWithType, contract],
+  )
+
+  const harvest = useCallback(
+    async (nftIds: BigNumber[], poolIds: BigNumber[]) => {
+      if (!contract) return
+
+      const encodeData = poolIds.map(id => defaultAbiCoder.encode(['tupple(uint256[] pIds)'], [{ pIds: [id] }]))
+
+      try {
+        const estimateGas = await contract.estimateGas.harvestMultiplePools(nftIds, encodeData)
+        const tx = await contract.harvestMultiplePools(nftIds, encodeData, {
+          gasLimit: calculateGasMargin(estimateGas),
+        })
+        addTransactionWithType(tx, { type: 'Harvest' })
+        return tx
+      } catch (e) {
+        console.log(e)
+      }
+    },
+    [addTransactionWithType, contract],
+  )
+
+  return { deposit, withdraw, approve, stake, unstake, harvest, emergencyWithdraw }
+}
+
+export const usePostionFilter = (positions: PositionDetails[], validPools: string[]) => {
+  const filterOptions = [
+    {
+      code: 'in_rage',
+      value: t`In range`,
+    },
+    {
+      code: 'out_range',
+      value: t`Out of range`,
+    },
+    {
+      code: 'all',
+      value: t`All positions`,
+    },
+  ]
+
+  const [activeFilter, setActiveFilter] = useState('all')
```

Comment:
```suggestion
  ] as const

  const [activeFilter, setActiveFilter] = useState<typeof filterOptions[number]['code']>('all')
```

Reply:
why do we need `as const` here?

Reply:
For typing
![image](https://user-images.githubusercontent.com/23650671/195011065-5cb06d2f-8829-4642-a880-42cf7b57c2b8.png)
![image](https://user-images.githubusercontent.com/23650671/195011111-43ac8fe8-3a60-43e3-b63c-70cedb61a5d7.png)

Reply:
therefore it makes `activeFilter` enum type `in_rage|out_range|all` instead of string type
![image](https://user-images.githubusercontent.com/23650671/195011220-0f206ffb-4fe1-424c-9d7f-6f67d872d7e5.png)
![image](https://user-images.githubusercontent.com/23650671/195011351-731d8770-fa92-41e2-a663-f53325247d71.png)

Reply:
make sense. updated